### ISSUE TITLE
Vulkan: 3D lights don't update their shadows if a static mesh is hidden or shown besides them

### ISSUE THREAD

**Issue by NedKJaims** [godotengine/godot#29856]
**Godot version:**
3.1.1 stable, 3.1.1 stable mono

**OS/device including version:**
Windows 10
OpenGL ES 3.0 Renderer: Intel(R) HD Graphics 4400
GLES3 and GLES2

**Issue description:**
1.-Directional Light does not work CullMask, even if you deactivate the first layer, it still affects the light to the objects that are in the layer (In every layer it does not work).
![m](https://user-images.githubusercontent.com/47610687/59633338-e1c89000-9111-11e9-9c74-bea11402d976.png)
![b](https://user-images.githubusercontent.com/47610687/59633345-e68d4400-9111-11e9-90e5-7254bb6ddf63.png)
2.-The shadows produced by SpotLight take time to update, for example an object changed the property from visible to false and this still produces shadow, until I move the SpotLight this updates the shadows.
![n](https://user-images.githubusercontent.com/47610687/59633592-73380200-9112-11e9-81cf-2232cd7f5239.png)

**Steps to reproduce:**

**Minimal reproduction project:**

**Comment by Calinou** [comment]
Cull mask in DirectionalLight isn't implemented yet, see #19438.

**Comment by NedKJaims** [comment]
Oh sorry I did not investigate it well, but the second one is a bug

**Comment by lawnjelly** [comment]
I've been getting this second bug too in current master, am investigating. My original situation was quite a bit more complex with cull masks but I think it's the same or similar bug. I've made a minimum reproduction project:

[LightCullBug.zip](https://github.com/godotengine/godot/files/3676625/LightCullBug.zip)

For future reference:
```
<lawnjelly> it's like something is getting lost when you hide, then reshow the light
<reduz> its odd though because that should not happen due to how it works
<lawnjelly> is there a a cache made of objects when you show it?
<reduz> yeah shadowmaps only update when things around it change
<lawnjelly> that get affected by the light
<lawnjelly> I'll try and work out where it does the caching
<reduz> its all in visual_server_scene.cpp
<lawnjelly> ok thanks
<reduz> check when it unpairs a light with an object
<reduz> it should make the light dirty
<reduz> for some reason this is not happening
<lawnjelly> ahh i see
```

**Comment by lawnjelly** [comment]
Note also that another 'workaround' for this bug until I get the best fix is to call `set_extra_cull_margin(0)` on an object after hiding or showing it. This forces the necessary updates to the lighting, but is less efficient (and more cumbersome) than the proper fix in core.

**Comment by akien-mga** [comment]
Fixed for the `3.2` branch with #32475, but a fix is still needed for `master`.

**Comment by clayjohn** [comment]
I can't reproduce in Beta 4. So closing this. Shadows have been worked on multiple times since this issue was made. I don't know when exactly it was fixed, but it is fixed now.
